sentinel-core/src/main/java/com/alibaba/csp/sentinel/slots/block/degrade/DegradeRuleManager.java

```diff
@@ -146,88 +142,128 @@ public static boolean setRulesForResource(String resourceName, Set<DegradeRule>
+                for (CircuitBreaker cb : e.getValue()) {
+                    rules.add(cb.getRule());
+                }
+                rm.put(e.getKey(), rules);
+            }
+
+            DegradeRuleManager.circuitBreakers = cbs;
+            DegradeRuleManager.ruleMap = rm;
+        }
+
         @Override
         public void configUpdate(List<DegradeRule> conf) {
-            Map<String, Set<DegradeRule>> rules = loadDegradeConf(conf);
-            if (rules != null) {
-                degradeRules.clear();
-                degradeRules.putAll(rules);
-            }
-            RecordLog.info("[DegradeRuleManager] Degrade rules received: " + degradeRules);
+            reloadFrom(conf);
+            RecordLog.info("[DegradeRuleManager] Degrade rules has been updated to: " + ruleMap);
         }
 
         @Override
         public void configLoad(List<DegradeRule> conf) {
-            Map<String, Set<DegradeRule>> rules = loadDegradeConf(conf);
-            if (rules != null) {
-                degradeRules.clear();
-                degradeRules.putAll(rules);
-            }
-            RecordLog.info("[DegradeRuleManager] Degrade rules loaded: " + degradeRules);
+            reloadFrom(conf);
+            RecordLog.info("[DegradeRuleManager] Degrade rules loaded: " + ruleMap);
         }
 
-        private Map<String, Set<DegradeRule>> loadDegradeConf(List<DegradeRule> list) {
-            Map<String, Set<DegradeRule>> newRuleMap = new ConcurrentHashMap<>();
-
+        private Map<String, List<CircuitBreaker>> buildCircuitBreakers(List<DegradeRule> list) {
+            Map<String, List<CircuitBreaker>> cbMap = new HashMap<>(8);
             if (list == null || list.isEmpty()) {
-                return newRuleMap;
+                return cbMap;
             }
-
             for (DegradeRule rule : list) {
                 if (!isValidRule(rule)) {
-                    RecordLog.warn(
-                        "[DegradeRuleManager] Ignoring invalid degrade rule when loading new rules: " + rule);
+                    RecordLog.warn("[DegradeRuleManager] Ignoring invalid rule when loading new rules: " + rule);
                     continue;
                 }
 
                 if (StringUtil.isBlank(rule.getLimitApp())) {
                     rule.setLimitApp(RuleConstant.LIMIT_APP_DEFAULT);
                 }
-
-                String identity = rule.getResource();
-                Set<DegradeRule> ruleSet = newRuleMap.get(identity);
-                if (ruleSet == null) {
-                    ruleSet = new HashSet<>();
-                    newRuleMap.put(identity, ruleSet);
+                CircuitBreaker cb = getExistingSameCbOrNew(rule);
+                if (cb == null) {
+                    RecordLog.warn("[DegradeRuleManager] Unknown circuit breaking strategy, ignoring: " + rule);
+                    continue;
                 }
-                ruleSet.add(rule);
```

Comment:
Inside `getExistingSameCbOrNew` method, the design of reuse the circuit breaker if the rule remains unchanged.is smart. I tried with `DegradeRuleManager.loadRules` in different places, it works. A small question is that, if using `DegradeRuleManager.loadRules(rules)` first time, and rules have two same `DegradeRule` , there will be two `CircuitBreaker` created, since in the first time, the static `circuitBreakers` variable is null.

Reply:
Yes, it can be improved.